### ISSUE TITLE
perf: updates dashboard slow

### ISSUE THREAD

**Issue by JakobLichterfeld** [teslamate-org/teslamate#4030]
Ok, I think we will go with query 8, will open a PR later. Any other queries to speed up? ,🤓

_Originally posted by @swiffer in https://github.com/teslamate-org/teslamate/issues/3921#issuecomment-2198258530_

---> The updates dashboard is slow @swiffer

**Comment by swiffer** [comment]
Could you run an explain analyze on the slow query and add it here? Guess my data is to small or the compute power to big 😀

**Comment by JakobLichterfeld** [comment]
:-D Yeah for sure. Just need to figure out the single query.

> or the compute power to big
Did not spot it on my Mac Studio M2 Pro, but on my Raspberry Pi 3 B and I know the updates are stored in separate table, therefore the time needed does not make sense to me.

![image](https://github.com/teslamate-org/teslamate/assets/16510328/476ee92c-882c-4249-aa5d-3184c56a5320)

**Comment by swiffer** [comment]
Took a Look already - it Scans positions and charges for 10 years. I guess this can be simplified without big sacrifice in data accuracy. Will come up with an alternative query, let's see how much they differ.

**Comment by swiffer** [comment]
@JakobLichterfeld - can you run this query (once with explain analyze and once without to compare data outcome)

```sql
with u as (

  select *, coalesce(lag(start_date) over(order by start_date desc), now()) as next_start_date 

  from updates

  where car_id = '2' and start_date BETWEEN '2014-07-04T12:37:24.227Z' AND '2024-07-04T12:37:24.227Z'

),

rng as (

  SELECT

	  date_trunc('hour', date) AS date,

	  (sum(rated_battery_range_km) / nullif(sum(usable_battery_level),0) * 100 ) AS "battery_rng",
	  
	  sum(case when action = 'Charge' then 1 else 0 end) as chg_ct

  FROM (

    select coalesce(usable_battery_level, battery_level) as usable_battery_level, start_date as date, start_rated_range_km as rated_battery_range_km, start_ideal_range_km as ideal_battery_range_km, 'Drive' as action

    from drives d
    
    inner join positions p on d.start_position_id = p.id 

    where d.car_id = '2' and d.start_date BETWEEN '2014-07-04T12:37:24.227Z' AND '2024-07-04T12:37:24.227Z'

    union all

    select end_battery_level as usable_battery_level, end_date, end_rated_range_km as rated_battery_range_km, end_ideal_range_km as ideal_battery_range_km, 'Charge' as action

    from charging_processes p

    where p.car_id = '2' and end_date BETWEEN '2014-07-04T12:37:24.227Z' AND '2024-07-04T12:37:24.227Z'

  ) as data

  GROUP BY 1

)


select	

  u.start_date as time,

	extract(EPOCH FROM u.end_date - u.start_date) AS update_duration,

	extract(EPOCH FROM u.start_date - lag(u.start_date) OVER (ORDER BY u.start_date)) AS since_last_update,

	split_part(u.version, ' ', 1) as version,

	sum(r.chg_ct) as chg_ct,

	convert_km(avg(r.battery_rng), 'km')::numeric(6,2) AS avg_rated_range_km

from u u

left join rng r

	ON r.date between u.start_date and u.next_start_date

group by u.car_id,

	u.start_date,

	u.end_date,

	next_start_date,

	split_part(u.version, ' ', 1)
```

**Comment by JakobLichterfeld** [comment]
> @JakobLichterfeld - can you run this query (once with explain analyze and once without to compare data outcome vs current dashboard result)

The query with `card_id = '1'`returns 136 rows for me and runs 420.269 ms

With EXPLAIN ANALYZE it takes 545.820 ms to return 26 rows ?
```
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 WindowAgg  (cost=17237.84..17242.26 rows=136 width=168) (actual time=543.859..545.200 rows=136 loops=1)
   ->  Sort  (cost=17237.84..17238.18 rows=136 width=122) (actual time=543.817..543.898 rows=136 loops=1)
         Sort Key: updates.start_date
         Sort Method: quicksort  Memory: 28kB
         ->  HashAggregate  (cost=17230.64..17233.02 rows=136 width=122) (actual time=542.824..543.554 rows=136 loops=1)
               Group Key: updates.car_id, updates.start_date, updates.end_date, (COALESCE((lag(updates.start_date) OVER (?))::timestamp with time zone, now())), split_part((updates.version)::text, ' '::text, 1)
               Batches: 1  Memory Usage: 88kB
               ->  Nested Loop Left Join  (cost=16643.72..17177.76 rows=3022 width=98) (actual time=144.089..534.579 rows=2005 loops=1)
                     Join Filter: ((r.date >= updates.start_date) AND (r.date <= (COALESCE((lag(updates.start_date) OVER (?))::timestamp with time zone, now()))))
                     Rows Removed by Join Filter: 273127
                     ->  WindowAgg  (cost=9.20..12.26 rows=136 width=53) (actual time=0.508..1.631 rows=136 loops=1)
                           ->  Sort  (cost=9.20..9.54 rows=136 width=41) (actual time=0.482..0.609 rows=136 loops=1)
                                 Sort Key: updates.start_date DESC
                                 Sort Method: quicksort  Memory: 26kB
                                 ->  Seq Scan on updates  (cost=0.00..4.38 rows=136 width=41) (actual time=0.045..0.244 rows=136 loops=1)
                                       Filter: ((start_date >= '2014-07-04 12:37:24.227'::timestamp without time zone) AND (start_date <= '2024-07-04 12:37:24.227'::timestamp without time zone) AND (car_id = '1'::smallint))
                     ->  Materialize  (cost=16634.52..16681.08 rows=200 width=48) (actual time=0.783..1.814 rows=2023 loops=136)
                           ->  Subquery Scan on r  (cost=16634.52..16680.08 rows=200 width=48) (actual time=106.474..136.080 rows=2023 loops=1)
                                 ->  GroupAggregate  (cost=16634.52..16678.08 rows=200 width=48) (actual time=106.470..133.886 rows=2023 loops=1)
                                       Group Key: (date_trunc('hour'::text, "*SELECT* 1".date))
                                       ->  Sort  (cost=16634.52..16640.95 rows=2571 width=48) (actual time=106.397..108.172 rows=2571 loops=1)
                                             Sort Key: (date_trunc('hour'::text, "*SELECT* 1".date))
                                             Sort Method: quicksort  Memory: 189kB
                                             ->  Result  (cost=0.43..16488.90 rows=2571 width=48) (actual time=0.088..97.473 rows=2571 loops=1)
                                                   ->  Append  (cost=0.43..16456.76 rows=2571 width=48) (actual time=0.078..85.684 rows=2571 loops=1)
                                                         ->  Subquery Scan on "*SELECT* 1"  (cost=0.43..16410.09 rows=1996 width=48) (actual time=0.076..79.200 rows=1996 loops=1)
                                                               ->  Nested Loop  (cost=0.43..16390.13 rows=1996 width=62) (actual time=0.073..76.010 rows=1996 loops=1)
                                                                     ->  Seq Scan on drives d  (cost=0.00..75.93 rows=1996 width=18) (actual time=0.017..7.368 rows=1996 loops=1)
                                                                           Filter: ((start_date >= '2014-07-04 12:37:24.227'::timestamp without time zone) AND (start_date <= '2024-07-04 12:37:24.227'::timestamp without time zone) AND (car_id = '1'::smallint))
                                                                     ->  Index Scan using positions_pkey on positions p  (cost=0.43..8.17 rows=1 width=8) (actual time=0.026..0.026 rows=1 loops=1996)
                                                                           Index Cond: (id = d.start_position_id)
                                                         ->  Subquery Scan on "*SELECT* 2"  (cost=0.00..33.81 rows=575 width=48) (actual time=0.081..3.813 rows=575 loops=1)
                                                               ->  Seq Scan on charging_processes p_1  (cost=0.00..28.06 rows=575 width=62) (actual time=0.076..2.772 rows=575 loops=1)
                                                                     Filter: ((end_date >= '2014-07-04 12:37:24.227'::timestamp without time zone) AND (end_date <= '2024-07-04 12:37:24.227'::timestamp without time zone) AND (car_id = '1'::smallint))
 Planning Time: 2.560 ms
 Execution Time: 545.820 ms
(36 rows)
```

**Comment by swiffer** [comment]
36 rows is the query plan, not the actual result set. Ok if you compare data compared by the query with the current dashboard how much does it differ? I'm expecting slight differences in the avg range and maybe number of charges column but as it's that much quicker I would say we use the updated one?

**Comment by JakobLichterfeld** [comment]
> but as it's that much quicker I would say we use the updated one?

That's my feeling as well. Will look more into detail by tomorrow. I do not see much benefit in number of charges and range anyways in the update dashboard.

**Comment by swiffer** [comment]
> I do not see much benefit in number of charges and range anyways in the update dashboard.

Same

**Comment by JakobLichterfeld** [comment]
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/jakob/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/jakob/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">


# Charges deviation (Dashboard - new query) | Range   deviation (Dashboard - new query)
-- | --
0 | 0,71
0 | 1,77
0 | -2,37
0 | -3,31
0 | -0,83
0 | -2,74
0 | 1,27
0 | 0,58
0 | 0,52
0 | -1,16
0 | 482,4
0 | 3,22
0 | 2,05
0 | 0,71
0 | 2,09
0 | 0,4
0 | 486,5
0 | 1,14
0 | 1,5
0 | -0,73
0 | 0,25
0 | 0,71
0 | 0,13
0 | 1,24
0 | 2,63
0 | 6,54
-1 | 0,59
1 | 3,69
0 | 5,2
0 | 3,67
0 | 2,82
0 | 3,58
0 | 3,19
0 | 4,6
0 | 4,54
0 | 4,44
0 | 3,64
0 | 4,77
0 | 13,79
0 | 6,04
0 | 2,75
0 | 3,39
-1 | 2,35
1 | 1,47
0 | 2,43
0 | 1,15
0 | 1,46
0 | 1,67
0 | 3,14
0 | 1,27
0 | 1,64
0 | 0,39
0 | 458,6
0 | 5,04
0 | 4,9
0 | 5,18
-1 | 4,89
1 | 15,32
0 | 3,77
0 | 6,33
0 | 3,77
0 | 5,04
-1 | 15,19
1 | 7,42
0 | 8,59
-1 | 7,79
1 | 3
0 | 4,53
0 | 4,19
-1 | 12,95
1 | 4,72
0 | 10,07
0 | 2,64
0 | 2,23
0 | 1,2
0 | 0,71
0 | 2,88
0 | -0,12
0 | 1,25
0 | 1,43
0 | 1,46
0 | 0,75
0 | 3,15
-1 | 1,51
1 | 2,5
-1 | 2,52
1 | 3,12
0 | 7,39
0 | 10,55
0 | 2,64
0 | 7,08
0 | 4,41
0 | 6,95
0 | 11,72
0 | 7,71
0 | 8,2
0 | 0,91
0 | 8,18
0 | 2,98
0 | 9,67
0 | 2,58
0 | 3,47
0 | 1,06
0 | 1,39
0 | -0,64
0 | 1,12
0 | 1,7
0 | 1,21
0 | -0,13
0 | 2,23
0 | 1,54
0 | 1,74
0 | 2,08
0 | 2,84
-1 | 3,13
1 | 6,32
0 | 5,44
0 | 6,11
0 | 6,22
0 | 14,57
0 | 9,22
0 | 0,56
0 | 5,17
0 | 6,58
0 | 15,53
0 | 8,83
0 | 7,58
0 | 8,5
0 | 1,39
0 | 2,03
0 | 1,41
0 | 1,61
0 | 0,06
0 | 0,95
0 | 0,6
0 | 445



</body>

</html>

**Comment by JakobLichterfeld** [comment]
closed by #4051

**Comment by JakobLichterfeld** [comment]
closed by #4051
